spec/app/ics-721-nft-transfer/README.md

```diff
@@ -0,0 +1,460 @@
+    } else { // we are sink chain
+      prefixedClassId = "{packet.destPort}/{packet.destChannel}/" + data.classId
+      // create NFT class if it doesn't exist already
+      if (nft.HasClass(prefixedClassId) === false) {
+        nft.SaveClass(data.classId, data.classUri)
+      }
+      // mint voucher to receiver
+      nft.Mint(prefixedClassId, data.tokenIds[i], data.tokenUris[i], data.receiver)
+    }
+  }
+}
+```
+
+`onAcknowledgePacket` is called by the routing module when a packet sent by this module has been acknowledged.
+
+```typescript
+function onAcknowledgePacket(
+  packet: Packet,
+  acknowledgement: bytes) {
+  // if the transfer failed, refund the tokens
+  if (!ack.success)
+    refundToken(packet)
+}
+```
+
+`onTimeoutPacket` is called by the routing module when a packet sent by this module has timed out (such that it will not be received on the destination chain).
+
+```typescript
+function onTimeoutPacket(packet: Packet) {
+  // the packet timed-out, so refund the tokens
+  refundToken(packet)
+}
+```
+
+`refundToken` is called by both `onAcknowledgePacket`, on failure, and `onTimeoutPacket`, to refund escrowed token to the original sender.
+
+```typescript
+function refundToken(packet: Packet) {
+  NonFungibleTokenPacketData data = packet.data
+  prefix = "{packet.sourcePort}/{packet.sourceChannel}/"
+  // we are the source if the classId is not prefixed with the packet's sourcePort and sourceChannel
+  source = data.classId.slice(0, len(prefix)) !== prefix
+  for (var i in data.tokenIds) { {
+    if source {
+      // unescrow token back to sender
+      nft.Transfer(data.classId, data.tokenIds[i], data.sender)
+    } else {
+      // we are sink chain, mint voucher back to sender
+      nft.Mint(data.classId, data.tokenIds[i], data.tokenUris[i], data.sender)
+    }
+  }
+}
+```
+
+```typescript
+function onTimeoutPacketClose(packet: Packet) {
+  // can't happen, only unordered channels allowed
+}
+```
+
+#### Reasoning
+
+##### Correctness
+
+This implementation preserves token non-fungibility and redeemability.
+
+* Non-fungibility: Only one instance of any token is *live* across all the IBC-connected blockchains.
+* Redeemability: If tokens have been sent to the counterparty chain, they can be redeemed back in the same `classId` & `tokenId` on the source chain.
+
+##### Multi-chain notes
+
+This specification does not directly handle the "diamond problem", where a user sends a token originating on chain A to chain B, then to chain D, and wants to return it through D -> C -> A — since the token is tracked as owned by chain B (and the `classId` will be "{portOnD}/{channelOnD}/{portOnB}/{channelOnB}/classId"), chain C cannot serve as the intermediary. The original path, in reverse order, is required to redeem the token at its source chain. Complexities arising from long redemption paths may lead to the emergence of central chains in the network topology.
```

Comment:
This text looks mostly lifted from the ICS20 spec. Better to remove it here and reference the ICS20 specs, as the problem is the same in both cases.